### DISCUSSION TITLE
Refresh Views

### DISCUSSION THREAD

**Discussion by hanan-commits-work** [iTwin/itwinjs-core#3826] [Q&A]
Is there any way to refresh views without reloading the entire app? 
In React, normally state variables and cache mechanism is used for this purpose. I was wondering if there is any built-in capability of iTwin that just refreshes the current view (events, reality models, zoom, etc remaining loaded).

**Comment by pmconne** [comment]
By "view" you mean "viewport", right? In most cases, a viewport is expected to automatically refresh itself whenever changes are made that require it to do so. Do you have an example of code that changes the viewport where the viewport doesn't automatically refresh? If so we can fix that. For example, if you are changing the planar clip mask settings for a reality model and the viewport is not updating to reflect the change, that sounds like a bug - show your code and we can fix it.

In some cases the viewport cannot detect that a change has occurred. It provides some methods you can call to tell it what changed, which will trigger a refresh on the next frame.
- [invalidateScene](https://www.itwinjs.org/reference/core-frontend/views/viewport/invalidatescene)
- [invalidateDecorations](https://www.itwinjs.org/reference/core-frontend/views/viewport/invalidatedecorations/)
- [setFeatureOverrideProviderChanged](https://www.itwinjs.org/reference/core-frontend/views/viewport/setfeatureoverrideproviderchanged/)

**Reply by hanan-commits-work** [reply]
Its not a change to the view exactly.
I simply want to create a button that refreshes the viewport.

**Comment by pmconne** [comment]
> I simply want to create a button that refreshes the viewport.

I don't understand why you would want that, or what you would expect it to do. If something changes about the view, the viewport should generally refresh automatically. If nothing changed about the view, refreshing the viewport should have zero effect.

If you are experiencing some problem that you think this button will solve, please describe the problem rather than the proposed solution.

**Reply by hanan-commits-work** [reply]
I didnt make it clear, apologies.
When we refresh the application, it goes back to the initial landing page. Login page in our case.
We want that when we press the refresh button, the user doesnt have to repeat the initial steps, such as Login, Selecting the Reality Model, or Showing the camera markers, etc. 
We want the state of the current user activity to be maintained even though the application is refreshed.

**Reply by pmconne** [reply]
Thanks for clarifying. Most of a viewport's state - like the models, categories, reality models, display style, camera position, etc - can be saved as and restored from [JSON](https://www.itwinjs.org/reference/core-common/views/viewdefinitionprops/). If you've added additional data like camera markers to your viewport, you'd be responsible for saving and restoring that data's state.Perhaps you can set up your app to listen for the page unload event and save that state somewhere to be reapplied to the viewport after the page reloads.

I can't help with the login screen or React stuff - @aruniverse or @kckst8 may be able to, if you provide information like what if any app template you're using etc.

**Reply by aruniverse** [reply]
Not sure why you have to login again, or what your code looks like for that. If youre using our BrowserAuthorizationClient, that should cache the token in local storage, not requiring to login again.

If you just copied the code from the template and havent decoupled the auth client from the viewer, then of course when you refresh the page all the components unmount and need to remount and reconstruct the auth client. You probably want to do a silent sign in if possible. We have plenty of examples like our [demo portal](https://github.com/iTwin/demo-portal/tree/master/packages/app) that you can use as an example. Refreshing the pg loses the view state context, but you don't have to sign in again
